src/material/schematics/ng-generate/mdc-migration/rules/template-migration.ts

```diff
@@ -10,27 +10,32 @@ import {Migration, ResolvedResource} from '@angular/cdk/schematics';
 import {SchematicContext} from '@angular-devkit/schematics';
 import {visitElements, parseTemplate} from './tree-traversal';
 import {ComponentMigrator} from '.';
+import {Update} from './template-migrator';
 
 export class TemplateMigration extends Migration<ComponentMigrator[], SchematicContext> {
   enabled = true;
 
   override visitTemplate(template: ResolvedResource) {
     const ast = parseTemplate(template.content, template.filePath);
     const migrators = this.upgradeData.filter(m => m.template).map(m => m.template!);
+    const updates: Update[] = [];
 
-    visitElements(
-      ast.nodes,
-      node => {
-        migrators.forEach(m => {
-          template.content = m.updateEndTag(template.content, node);
-        });
-      },
-      node => {
-        migrators.forEach(m => {
-          template.content = m.updateStartTag(template.content, node);
-        });
-      },
-    );
+    visitElements(ast.nodes, node => {
+      for (let i = 0; i < migrators.length; i++) {
+        updates.push(...migrators[i].getUpdates(node));
+      }
+    });
+
+    updates.sort((a, b) => {
+      if (a.location.line !== b.location.line) {
+        return b.location.line - a.location.line;
+      }
+      return b.location.col - a.location.col;
```

Comment:
Could you just use the offset/index? (instead of having to deal with line and column)

Reply:
Yep. I'm actually sad that I didn't see that solution myself T-T